safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,267 @@
+    let mut wal_seg_size = usize::try_from(wal_seg_size).with_context(|| format!("Couldn't convert {} to usize", wal_seg_size))?;
+    if !(MIN_WAL_SEGMENT_SIZE..=MAX_WAL_SEGMENT_SIZE).contains(&wal_seg_size) {
+        warn!("Invalid wal seg size provided, should be at least 1Mb per postgres, setting to default value of {}", DEFAULT_WAL_SEGMENT_SIZE);
+        wal_seg_size = DEFAULT_WAL_SEGMENT_SIZE; 
+    }
+
+    let mut backup_lsn = backup_start;
+
+    loop {
+        if let Some(lease_id) = lease_id {
+            broker::become_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.ok();
+        }
+
+        while lsn_durable.changed().await.is_ok() {
+            let commit_lsn = *lsn_durable.borrow();
+
+            if let Some(lease_id) = lease_id {
+                if broker::is_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.is_err() {
+                    info!("Leadership is lost for timeline {}", timeline_id);
+                    break;
+                }
+            }
+
+            if commit_lsn < backup_lsn {
+                bail!("Commit lsn {} moved back {}", commit_lsn, backup_lsn);
+            }
+
+            if backup_lsn.segment_number(wal_seg_size) == commit_lsn.segment_number(wal_seg_size) {
+                continue;
+            }
+
+            info!("Woken up for lsn {} committed, will back it up. backup lsn {}", commit_lsn, backup_lsn);
+
+            backup_lsn = backup_lsn_range(backup_lsn, commit_lsn, wal_seg_size, &conf, timeline_id).await?;
+            lsn_backed_up.send(backup_lsn)?;
+        }
+    }
+}
+
+pub async fn backup_lsn_range(start_lsn: Lsn, end_lsn: Lsn, wal_seg_size: usize, conf: &SafeKeeperConf, timeline_id: ZTenantTimelineId) -> Result<Lsn> {
+
+    let mut res = start_lsn;
+    for s in get_segments(start_lsn, end_lsn, wal_seg_size) {
+
+        let seg_backup = backup_single_segment(s, wal_seg_size, conf, timeline_id).await;
+
+        // TODO: limit this only to Not Found errors
+        if seg_backup.is_err() && start_lsn.is_valid() {
+            error!("Segment {} not found in timeline {}", s.seg_no, timeline_id)
+        }
+
+        ensure!(start_lsn >= s.start_lsn || start_lsn.is_valid(), "Out of order segment upload detected");
+
+        if res == s.start_lsn {
+            res = s.end_lsn;
+        } else {
+            warn!("Out of order Segment {} upload had been detected for timeline {}. Backup Lsn {}, Segment Start Lsn {}", s.seg_no, timeline_id, res, s.start_lsn)
+        }
+    }
+
+    Ok(res)
+}
+
+async fn backup_single_segment(seg: Segment, wal_seg_size: usize, conf: &SafeKeeperConf, timeline_id: ZTenantTimelineId) -> Result<()> {
+
+    let segment_file_name = seg.file_path(&timeline_id, conf)?;
+    let dest_name = PathBuf::from(format!("{}/{}", timeline_id.tenant_id, timeline_id.timeline_id)).with_file_name(seg.file_name());
+
+    info!("Backup of {:?} requested", segment_file_name);
```

Comment:
Styling nit:

```suggestion
    info!("Backup of {} requested", segment_file_name.display());
```

A bunch more plaves can have this changed.

Reply:
applied